Add Image Block 'Link to' setting

talldan opened the pull request:
## Description
closes #5309 

Adds a new 'Link To' setting in the image block settings, similar to the 'Link To' option in the old editor.

Removes the old url editor toolbar option for the image block (this could be kept, though would need a bit of thought about how the two could co-exist)

## How has this been tested?
- Manual testing. Tested each of the settings individually. Tested converting from a classic editor image to ensure the link url is retained in the image block. 
- Added 'full-content' tests for various settings of the new `linkDestination` attribute

## Screenshots <!-- if applicable -->
![link-to-update](https://user-images.githubusercontent.com/677833/42261113-a2e96fb0-7f99-11e8-9b91-8917aba76980.gif)

## Types of changes
Backwards compatibility with classic editor images.
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

talldan committed:
Add a new image block setting for 'Link to'

- Add a new SelectControl to the InspectorControls
- The options for this are sourced from a new function,
`getLinkDestinationOptions`
- `getLinkDestinationOptions` returns options for the mediaFile or the
attachment page
- Custom url to be added in a later commit

talldan committed:
Store value of 'Link To' dropdown as separate attribute

- Introduce new attribute `linkDestination` for storing the selected
option of the 'Link To' block setting
- Use constants for the value of this dropdown
- Derive href based on value of dropdown
- Add a text field for entering a custom url

talldan committed:
Change order of Link To select options to reflect None as the default

talldan committed:
Add snapshot test for image block editor

talldan commented on core-blocks/image/edit.js:
```diff
@@ -52,6 +51,13 @@ import ImageSize from './image-size';
 const MIN_SIZE = 20;
 
+const LINK_DESTINATION_OPTIONS = {
+	NONE: 'NONE',
```
Unsure about whether to use all caps for these constant properties. Could also make them individual consts instead of putting them in an object.

noisysocks replied:
> Could also make them individual consts instead of putting them in an object.

I'd go with this 🙂

talldan commented:
@karmatosed Would be great to get some design feedback on this one if you get a moment. 

In particular I'm wondering if the input field for entering a custom url could be improved.

This is what I've done:
![link-to-dropdown](https://user-images.githubusercontent.com/677833/42209110-6c9ab3f6-7ee0-11e8-9639-105ccdecb2b7.gif)

Here's a gif of the classic editor for contrast:
![link-to-classic](https://user-images.githubusercontent.com/677833/42212925-0d9c10ac-7eea-11e8-8977-436a22b6ff89.gif)

karmatosed reviewed (changes requested):
@talldan I think if we have a disabled state on first load of the input that would make more sense. As it's a guide not probably what someone will have? Then you can click in and it changes to active input state?

noisysocks commented on core-blocks/image/test/index.js:
```diff
@@ -0,0 +1,13 @@
+	test( 'block edit matches snapshot', () => {
+		const wrapper = blockEditRender( name, settings );
+
+		expect( wrapper ).toMatchSnapshot();
```
I think this is because `blockEditRender` skips the necessary bootstrapping for `select()` to work, namely `wp.data.registerStore()`. We could export a non-enhanced version of the edit component for testing. Or, we could just not bother 🙂

talldan committed:
Revert "Add snapshot test for image block editor"

This reverts commit 4d707a31606c8a79d96a20b8bb68555658a1f9d7.

Reverted for now as blockEditRender doesn't seem to support select.

talldan committed:
Always render the 'Custom URL' input field

This change makes the Custom URL field always visible, but renders it as
disabled unless the 'Custom URL' option is selected in the Link to
dropdown.

talldan committed:
Usability changes for 'Custom Url' input field in image block settings

Make the placeholder only display when the image block is editable (not
disabled)

talldan committed:
Move 'Link To' and 'Link URL' image block settings to their own PanelBody

This makes it a bit more consistent with the File Block's Link To setting.

Have also changed the Custom URL field label to Link URL.

talldan committed:
Change link destination constants to individual variables

talldan committed:
Handle setting of image block href when converting from classic to blocks

The strategy chosen is to always set 'Link To' to 'Custom URL' if an anchor tag
exists in the classic markup. Trying to detect 'media file' or
'attachment page' could potentially be too error prone, this approach at least
ensures the image's anchor link is the same in the resulting post.

talldan commented:
@karmatosed I've made some changes along those lines. Since the input field with the url in it will always be visible, I've moved the two fields into their own section in the settings so that it feels contextually better. That feels a lot tidier to me and is also consistent with the new File Block. Here's a gif, let me know what you think:

![link-to-update](https://user-images.githubusercontent.com/677833/42261113-a2e96fb0-7f99-11e8-9b91-8917aba76980.gif)

talldan committed:
linkDestination=undefined when transforming a raw image block with no <a> tag

For the raw transform on an image block, set the linkDestination to undefined
if the raw html does not contain an anchor tag (or has an anchor with no href).
This more accurately reflects that the default value of linkDestination is
undefined

talldan committed:
Make placeholder https instead of http, ready for the modern web

karmatosed reviewed (approved):
This looks good to me now. I am not totally happy with the way the input looks but that's using the default one so not something to fix in this PR. Let's get it in! Thanks for all your hardwork on this.

tofumatt commented on core-blocks/image/edit.js:
```diff
@@ -308,6 +339,21 @@ class ImageEdit extends Component {
+						label={ __( 'Link URL' ) }
+						value={ href }
+						onChange={ this.onSetCustomHref }
+						placeholder={ ! isLinkUrlInputDisabled && 'https://' }
```
Subtle hint toward an encrypted web, eh? 😄

tofumatt commented on core-blocks/image/index.js:
```diff
@@ -113,8 +116,11 @@ export const settings = {
 					const align = alignMatches ? alignMatches[ 1 ] : undefined;
 					const idMatches = /(?:^|\s)wp-image-(\d+)(?:$|\s)/.exec( className );
 					const id = idMatches ? idMatches[ 1 ] : undefined;
+					const anchorEl = node.querySelector( 'a' );
```
Can you rename this `anchorElement` instead of the abbreviation? Two-letter abbreviations are somewhat cryptic and unnecessry 😄

tofumatt commented on core-blocks/image/edit.js:
```diff
@@ -122,7 +126,26 @@ class ImageEdit extends Component {
+		let href;
+
+		if ( value === LINK_DESTINATION_NONE ) {
+			href = '';
```
Is there a reason we want this to be an empty string instead of `null`? `null` seems more appropriate.

Also, I'd just set the "none" value as the default here; start with `let href = '';` (or `let href = null;` and then change it if the link destination is something else.

noisysocks replied:
Adding `default: 'none'` to the attribute definition itself feels right to me in this case.

tofumatt replied:
Yeah that's fine too. Just something over the way it's being done now would be good.

talldan replied:
I've added `default: 'none'` 👍 

Also added some extra test fixtures to the full-content test, so that's a little bit of coverage.

tofumatt reviewed (approved):
I think there are a few minor things to address here first… also it would be nice if we could add tests for this, but I see there aren't any existing ones so perhaps that's a rabbit hole. 😒

talldan committed:
Rename `anchorEl` to `anchorElement`

talldan committed:
Make linkDestination attribute have default value of 'none'

talldan committed:
Add 'full-content' tests for various linkDestination values

talldan committed:
Change `href` to `undefined` when `linkDestination` of `none` is selected

For this to work, have had to use `value={ url || '' }` on the TextControl
since only an empty string can clear a controlled react input.
